### PR TITLE
Add a PWM section in the i.MX porting guide

### PR DESCRIPTION
**Pull request by MHesham**

### CODE REVIEW

File: Documentation/porting-imx.md [hunk 1]
```diff
@@ -410,6 +410,20 @@ Edit `Dsdt-Sdhc.inc` and ensure that the SDHC instance on which your boot media
...
+For example on setting the PWM deivce interface properties statically from an inf file, see the [Virtual PWM driver sample here](https://github.com/Microsoft/Windows-iotcore-samples/tree/develop/Drivers).
+
+For example on how to read ACPI _DSD in general from within a WDF driver and assigning so you can assign the PWM device properties in runtime, see the [i.MX SDHC driver here](https://github.com/MHesham/imx-iotcore/tree/public_preview/driver/sd/imxusdhc).
+
+#### PWM References:
+
+- [PWM DDI](https://docs.microsoft.com/en-us/windows-hardware/drivers/spb/pulse-width-controller%20driver).
+- [PWM Driver Reference](https://docs.microsoft.com/en-us/windows/desktop/devio/pwm-api).
+- [PWM WinRT APIs](https://docs.microsoft.com/en-us/uwp/api/windows.devices.pwm).
+- [Virtual PWM Driver Sample](https://github.com/Microsoft/Windows-iotcore-samples/tree/develop/Drivers).
```
Comment (hbeberman) [inline Documentation/porting-imx.md]: No need for periods after these links. They aren't sentences.

File: Documentation/porting-imx.md [hunk 2]
```diff
@@ -410,6 +410,20 @@ Edit `Dsdt-Sdhc.inc` and ensure that the SDHC instance on which your boot media
 ```c
 Package (2) { "Regulator1V8Exist", 0 }, // 1.8V Switching External Circuitry: Not implemented
 ```
+### PWM
+
+For best experience using the PWM WinRT APIs from UWP apps, some device properties may need to be set where some of which are generally required for UWP access to the PWM device interface while others are required for some PWM WinRT APIs to function correctly. More on these device interface properties can be found here in the [Setting device interface section](https://docs.microsoft.com/en-us/windows-hardware/drivers/spb/pulse-width-controller%20driver#setting-device-interface-properties) of the PWM DDI MSDN documentation.
+
+For example on setting the PWM deivce interface properties statically from an inf file, see the [Virtual PWM driver sample here](https://github.com/Microsoft/Windows-iotcore-samples/tree/develop/Drivers).
+
+For example on how to read ACPI _DSD in general from within a WDF driver and assigning so you can assign the PWM device properties in runtime, see the [i.MX SDHC driver here](https://github.com/MHesham/imx-iotcore/tree/public_preview/driver/sd/imxusdhc).
```
Comment (hbeberman) [inline Documentation/porting-imx.md]: This should be a relative link to /driver/sd/imxusdhc, not an absolute link to your fork.

File: Documentation/porting-imx.md [hunk 3]
```diff
@@ -410,6 +410,20 @@ Edit `Dsdt-Sdhc.inc` and ensure that the SDHC instance on which your boot media
 ```c
 Package (2) { "Regulator1V8Exist", 0 }, // 1.8V Switching External Circuitry: Not implemented
 ```
+### PWM
+
+For best experience using the PWM WinRT APIs from UWP apps, some device properties may need to be set where some of which are generally required for UWP access to the PWM device interface while others are required for some PWM WinRT APIs to function correctly. More on these device interface properties can be found here in the [Setting device interface section](https://docs.microsoft.com/en-us/windows-hardware/drivers/spb/pulse-width-controller%20driver#setting-device-interface-properties) of the PWM DDI MSDN documentation.
+
+For example on setting the PWM deivce interface properties statically from an inf file, see the [Virtual PWM driver sample here](https://github.com/Microsoft/Windows-iotcore-samples/tree/develop/Drivers).
```
Comment (hbeberman) [inline Documentation/porting-imx.md]:
```suggestion
For an example of setting the PWM deivce interface properties statically from an inf file, see the [Virtual PWM driver sample](https://github.com/Microsoft/Windows-iotcore-samples/tree/develop/Drivers).
```

Comment (hbeberman) [reply Documentation/porting-imx.md]: We shouldn't merge this pull until the virtual PWM driver has been checked into Windows-iotcore-samples/Drivers, and this URL should be updated directly to its folder.

File: Documentation/porting-imx.md (hunk 2 above)
Comment (hbeberman) [reply Documentation/porting-imx.md]: @christopherco Didn't we remove the _DSD dependency from the imxusdhc stack?

https://github.com/ms-iot/imx-edk2-platforms/blob/imx/Platform/SolidRun/HummingBoardEdge_iMX6Q_2GB/AcpiTables/Dsdt-Sdhc.asl

Comment (hbeberman) [inline Documentation/porting-imx.md]:
```suggestion
For an example of how to read the ACPI _DSD from within a WDF driver and how you can assign the PWM device properties at runtime, see the [i.MX SDHC driver here](https://github.com/MHesham/imx-iotcore/tree/public_preview/driver/sd/imxusdhc).
```

Comment (hbeberman) [inline Documentation/porting-imx.md]:
```suggestion
For an example of how to read the ACPI _DSD from within a WDF driver and how you can assign the PWM device properties at runtime, see the [i.MX SDHC driver here](https://github.com/MHesham/imx-iotcore/tree/public_preview/driver/sd/imxusdhc).
```
How does the SDHC driver relate to assigning PWM device properties at runtime?

File: Documentation/porting-imx.md [hunk 4]
```diff
@@ -410,6 +410,20 @@ Edit `Dsdt-Sdhc.inc` and ensure that the SDHC instance on which your boot media
 ```c
 Package (2) { "Regulator1V8Exist", 0 }, // 1.8V Switching External Circuitry: Not implemented
 ```
+### PWM
+
+For best experience using the PWM WinRT APIs from UWP apps, some device properties may need to be set where some of which are generally required for UWP access to the PWM device interface while others are required for some PWM WinRT APIs to function correctly. More on these device interface properties can be found here in the [Setting device interface section](https://docs.microsoft.com/en-us/windows-hardware/drivers/spb/pulse-width-controller%20driver#setting-device-interface-properties) of the PWM DDI MSDN documentation.
```
Comment (hbeberman) [inline Documentation/porting-imx.md]:
```suggestion
For the best experience using the PWM WinRT APIs from UWP apps some additional device properties need to be set. Documentation on these device interface properties can be found here in the [Setting device interface properties](https://docs.microsoft.com/en-us/windows-hardware/drivers/spb/pulse-width-controller%20driver#setting-device-interface-properties) section of the PWM DDI MSDN article.
```

**Review by hbeberman — CHANGES_REQUESTED** [review]
Please see the individual comments and suggestions.

File: Documentation/porting-imx.md (hunk 2 above)
Comment (hbeberman) [inline Documentation/porting-imx.md]: @christopherco Didn't we remove the _DSD dependency from the imxusdhc stack?

https://github.com/ms-iot/imx-edk2-platforms/blob/imx/Platform/SolidRun/HummingBoardEdge_iMX6Q_2GB/AcpiTables/Dsdt-Sdhc.asl

Comment (MHesham) [reply Documentation/porting-imx.md]: We did, but the code that deals with ACPI _DSD and _DSM is still there in the usdhc code, but it is dormant at the moment. Should be enabled again once we fix the SD power issues. Other aspects of the _DSD/_DSM needs some rework to be standard compliant, but the acpiutil code is still a good reference.

Comment (MHesham) [reply Documentation/porting-imx.md]: You are correct, the wording was wrong.

File: Documentation/porting-imx.md [hunk 5]
```diff
@@ -410,6 +410,20 @@ Edit `Dsdt-Sdhc.inc` and ensure that the SDHC instance on which your boot media
 ```c
 Package (2) { "Regulator1V8Exist", 0 }, // 1.8V Switching External Circuitry: Not implemented
 ```
+### PWM
+
+For the best experience using the PWM WinRT APIs from UWP apps some additional device properties need to be set. Documentation on these device interface properties can be found here in the [Setting device interface properties](https://docs.microsoft.com/en-us/windows-hardware/drivers/spb/pulse-width-controller%20driver#setting-device-interface-properties) section of the PWM DDI MSDN article.
+
+For an example of setting the PWM deivce interface properties statically from an inf file, see the [Virtual PWM driver sample [Virtual PWM driver sample here](https://github.com/Microsoft/Windows-iotcore-samples/tree/develop/Drivers/VirtualPWM).
+
+For an example of how to read the ACPI _DSD from within a WDF driver and how you can assign the PWM device properties at runtime, see the [i.MX SDHC driver here](https://github.com/MHesham/imx-iotcore/tree/public_preview/driver/sd/imxusdhc).
```
Comment (hbeberman) [inline Documentation/porting-imx.md]:
```suggestion
For an example of how to read the ACPI _DSD from within a WDF driver and how you can assign the PWM device properties at runtime, see the [i.MX SDHC driver here](../driver/sd/imxusdhc).
```

File: Documentation/porting-imx.md (hunk 2 above)
Comment (christopherco) [reply Documentation/porting-imx.md]: The problem with any sort of _DSD/_DSM we have previously implemented is that the definitions/behaviors are not formally documented in a specification (Github readmes or code comments do not count). Upstream edk2 requires references a public documents for these types of OS-UEFI interactions.

File: Documentation/porting-imx.md [hunk 6]
```diff
@@ -410,6 +410,20 @@ Edit `Dsdt-Sdhc.inc` and ensure that the SDHC instance on which your boot media
 ```c
 Package (2) { "Regulator1V8Exist", 0 }, // 1.8V Switching External Circuitry: Not implemented
 ```
+### PWM
+
+For the best experience using the PWM WinRT APIs from UWP apps some additional device properties need to be set. Documentation on these device interface properties can be found here in the [Setting device interface properties](https://docs.microsoft.com/en-us/windows-hardware/drivers/spb/pulse-width-controller%20driver#setting-device-interface-properties) section of the PWM DDI MSDN article.
+
+For an example of setting the PWM deivce interface properties statically from an inf file, see the [Virtual PWM driver sample [Virtual PWM driver sample here](https://github.com/Microsoft/Windows-iotcore-samples/tree/develop/Drivers/VirtualPWM).
```
Comment (christopherco) [inline Documentation/porting-imx.md]:
```suggestion
For an example of setting the PWM deivce interface properties statically from an inf file, see the [Virtual PWM driver sample](https://github.com/Microsoft/Windows-iotcore-samples/tree/develop/Drivers/VirtualPWM).
```

**Review by christopherco — APPROVED** [review]

**Review by hbeberman — APPROVED** [review]

**Pull request merged**